File: 백준/Gold/2608. 로마 숫자/로마 숫자.py
num1 = input()
num2 = input()

num_dict = {'M' : 1000, 'CM' : 900, 'D' : 500, 'CD' : 400, 'C':100, 'XC':90, 'L':50, 'XL':40, 'X':10, 'IX':9, 'V':5, 'IV':4, 'I':1}
char_dict = {1000:'M', 900 :'CM', 500:'D', 400:'CD', 100:'C', 90:'XC', 50:'L', 40:'XL', 10:'X', 9:'IX', 5:'V', 4:'IV', 1:'I'}

check = ['IV', 'IX', 'XL', 'XC', 'CD', 'CM']

N = len(num1)
idx = 0
n1 = 0
while idx < N:
    if idx < N-1:
        tmp = num1[idx] + num1[idx+1]
        if tmp in check:
            n1 += num_dict[tmp]
            idx += 2
        else:
            n1 += num_dict[num1[idx]]
            idx += 1
    else:
        n1 += num_dict[num1[idx]]
        idx += 1

N = len(num2)
idx = 0
n2 = 0
while idx < N:
    if idx < N-1:
        tmp = num2[idx] + num2[idx+1]
        if tmp in check:
            n2 += num_dict[tmp]
            idx += 2
        else:
            n2 += num_dict[num2[idx]]
            idx += 1
    else:
        n2 += num_dict[num2[idx]]
        idx += 1
# print(n1, n2)
total = n1 + n2
print(total)

answer = ''

nums = list(char_dict.keys())
nums.sort(reverse=True)


char_cnt = {'V':0, 'L':0, 'D':0, 'IV':0, 'IX':0, 'XL':0, 'XC':0, 'CD':0, 'CM':0}
for k in nums:
    cnt = 0
    c = char_dict[k]
    while total >= k:
        if cnt >= 3:
            break

        if c in char_cnt and char_cnt[c] > 0:
            break

        if c == 'IV' and answer[-2:] == 'IX':
            break
        elif c == 'XL' and answer[-2:] == 'XC':
            break
        elif c == 'CM' and answer[-2:] == 'CD':
            break

        total -= k
        answer += c
        if c in char_cnt:
            char_cnt[c] += 1
        else:
            cnt += 1

print(answer)

File: 백준/Gold/2608. 로마 숫자/로마 숫자.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String num1 = br.readLine();
        String num2 = br.readLine();

        int[] nums = new int[] {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] chars = new String[] {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
        Map<Integer, String> numMap = new HashMap<>();
        Map<String, Integer> charMap = new HashMap<>();
        Set<String> check = new HashSet<>(Arrays.asList("V", "L", "D", "IV", "IX", "XL", "XC", "CD", "CM"));
        for (int i = 0; i < nums.length; i++) {
            numMap.put(nums[i], chars[i]);
            charMap.put(chars[i], nums[i]);
        }

        int n1 = 0;
        int n = num1.length();
        int idx = 0;
        while (idx < n) {
            String p = String.valueOf(num1.charAt(idx));
            if (idx < n-1) {
                String tmp = num1.substring(idx, idx+2);
                if (check.contains(tmp)) {
                    n1 += charMap.get(tmp);
                    idx += 2;
                } else {
                    n1 += charMap.get(p);
                    idx += 1;
                }
            } else {
                n1 += charMap.get(p);
                idx += 1;
            }
        }

        int n2 = 0;
        n = num2.length();
        idx = 0;
        while (idx < n) {
            String p = String.valueOf(num2.charAt(idx));
            if (idx < n-1) {
                String tmp = num2.substring(idx, idx+2);
                if (check.contains(tmp)) {
                    n2 += charMap.get(tmp);
                    idx += 2;
                } else {
                    n2 += charMap.get(p);
                    idx += 1;
                }
            } else {
                n2 += charMap.get(p);
                idx += 1;
            }
        }

        int total = n1 + n2;
        System.out.println(total);

        String answer = "";
        Set<String> used = new HashSet<>();
        for (int k : nums) {
            int cnt = 0;
            String target = numMap.get(k);
            while (total >= k) {
                if (check.contains(target) && used.contains(target)) break;

                if (cnt >= 3) break;

                n = answer.length();
                if (n >= 2) {
                    if (target.equals("IV") && answer.substring(n-2, n).equals("IX")) {
                        break;
                    } else if (target.equals("XL") && answer.substring(n-2, n).equals("XC")) {
                        break;
                    } else if (target.equals("CM") && answer.substring(n-2, n).equals("CD")) {
                        break;
                    }
                }


                total -= k;
                answer += "" + target;
                used.add(target);
                cnt++;
            }
        }

        System.out.print(answer);
    }
}
